### DISCUSSION TITLE
How to get the real consumption reporting data in Media Player

### DISCUSSION THREAD

**Discussion by shilinding** [5G-MAG/rt-5gms-media-stream-handler#50] [General]
# Problem description
TS26.512-h40 clause 4.7.4 only defines the conditions for consumption reporting, but I can't find how to get the real consumption reporting data in the Media Stream Handler. 

The conditions are defined as follows:

> If the consumption reporting procedure is activated, the Media Session Handler shall submit a consumption report to the 5GMSd AF when any of the following conditions occur:
> -	Start of consumption of a downlink streaming session;
> -	Stop of consumption of a downlink streaming session;
> -	Upon determining the need to report ongoing 5GMS consumption at periodic intervals determined by the `clientConsumptionReportingConfiguration.reportingInterval` property.
> -	Upon determining a location change, if the `clientConsumptionReportingConfiguration.locationReporting` property is set to True.
> -	Upon determining an access network change (e.g. unicast to eMBMS, or vice versa), if the `clientConsumptionReportingConfiguration.accessReporting` property is set to True.

**Comment by rjb1000** [comment]
@shilinding: Is this a question about the client API between the Media Stream Handler (Media Player) and the Media Session Handler? Have you checked clause 13 of TS 26.512 which describes the **Media Stream Handler API** at reference point M7?

Relevant to your question, I see the following general statement in clause 13.2.7:
> The Media Session Handler may use the notifications, errors and status information provided through M7d to execute relevant tasks.

More specifically:
* Clause 13.2.5 defines **notification events** that the Media Session Handler can subscribe to which inform it about the current status of media playback.
  * The Media Session Handler can subscribe to these notification events and use them to determine _when_ to send consumption reports to the 5GMSd AF.
* When it receives a notification event from the Media Stream Handler (Media Player), the Media Session Handler can query the **dynamic status information** of the Media Stream Handler (Media Player) via reference point M7 to obtain additional information, as specified in clause 13.2.6.

Combined with information about each current media session latched by the Media Session Handler at the start of each media streaming session (e.g. the Media Entry Point URL), and other information gathered through local Android APIs, does this provide you with the information you need in order to populate a Consumption Report?

It feels like a short piece of analysis would help bring this problem into better focus. Perhaps you could create a little HTML table below listing the fields you need to populate and the source of the information to populate each field. (This would also be a good documentation of your implementation design.)

**Comment by rjb1000** [comment]
To detect a change of UE location (one of the consumption reporting triggers you list above), the Media Session Handler may need to use a suitable Android API outside the scope of 3GPP standardisation.

For the timer-based consumption reporting trigger you will need to find a suitable Android timer API, of course.

**Comment by shilinding** [comment]
Thanks Richard.
Sorry, maybe I didn't make my question clear. My main question is, how can we get the  consumption reporting in Media Stream Handler when the above reporting conditions triggered? Because I don't know how to get consumption reporting, now I'm just constructing fake data for testing.

**Comment by rjb1000** [comment]
> My main question is, how can we get the consumption reporting in Media Stream Handler when the above reporting conditions triggered? Because I don't know how to get consumption reporting, now I'm just constructing fake data for testing.

Ah. I see. So you question is about how the Media Stream Handler gets the required information from its Media Player (e.g. Exoplayer)?

If I have understood correctly, this is not feedback on 3GPP standards, but rather an implementation question. So I think this issue has been raised in the wrong repository.

Suggest you close this issue and open a new one in here: https://github.com/5G-MAG/rt-5gms-media-session-handler/issues/, tagging it as a _Question_ and assigning to @dsilhavy.

**Comment by dsilhavy** [comment]
* 26512 11.3.3.1 defines the consumption report format. It outlines what needs to be reported. Basically, it is the consumed `Representation@id` for DASH. We probably even have some overlaps here to metric reporting and the `RepSwitchEventType`
* In terms of collecting the required information you need to check the APIs of the Exoplayer. As an example, there is [`onTracksChanged`](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/analytics/AnalyticsListener.html#onTracksChanged(com.google.android.exoplayer2.analytics.AnalyticsListener.EventTime,com.google.android.exoplayer2.Tracks)) event that you can register for when defining an [AnalyticsListener](https://developer.android.com/guide/topics/media/exoplayer/analytics). See also here: https://exoplayer.dev/track-selection.html

@shilinding I suggest you look into the ExoPlayer API to find the right calls. As Richard suggested it might be good to create a simple table like this:

| Required Information | ExoPlayer API / How to get the information |
| ------------- | ------------- |
| mediaConsumed  | player. getCurrentTracks() or event  |

I will do something similar for metric reporting. Maybe this also helps: https://exoplayer.dev/track-selection.html

**Comment by haudiobe** [comment]
@dsilhavy why did you move this one away from the standards repo?

**Comment by dsilhavy** [comment]
> @dsilhavy why did you move this one away from the standards repo?

Because from my understanding, the questions are not specification related but rather how to derive the required information from the media player
